rimage/transform/parallel_projection.go

```diff
@@ -95,5 +97,214 @@ func (pp *ParallelProjection) PointCloudToRGBD(cloud pointcloud.PointCloud) (*ri
+// a shade of blue if it is below the miss threshold. These shades will be more distinct the further from the threshold they are.
+func getColorFromProbabilityValue(d pointcloud.Data) (rimage.Color, error) {
+	var r, g, b uint8
+
+	if d == nil {
+		return rimage.NewColor(0, 0, 0), errors.New("data received was null")
+	}
+
+	if !d.HasValue() {
+		return rimage.NewColor(255, 255, 255), nil
+	}
+
+	if d.Value() > 100 || d.Value() < 0 {
+		return rimage.NewColor(0, 0, 0),
+			errors.Errorf("received a value of %v which is outside the range (0 - 100) representing probabilities", d.Value())
+	}
+
+	prob := float64(d.Value()) / 100.
+
+	switch {
+	case prob < missThreshold:
+		b = uint8(255 * ((missThreshold - prob) / (hitThreshold - 0)))
+	case prob > hitThreshold:
+		g = uint8(255 * ((prob - hitThreshold) / (1 - missThreshold)))
+	default:
+		b = uint8(255 * ((missThreshold - prob) / (hitThreshold - 0)))
+		g = uint8(255 * ((prob - hitThreshold) / (1 - missThreshold)))
+	}
+
+	return rimage.NewColor(r, g, b), nil
+}
+
+// NewParallelProjectionOntoXZWithRobotMarker creates a new ParallelProjectionOntoXZWithRobotMarker with the given
+// robot pose.
+func NewParallelProjectionOntoXZWithRobotMarker(rp *spatialmath.Pose) ParallelProjectionOntoXZWithRobotMarker {
+	return ParallelProjectionOntoXZWithRobotMarker{robotPose: rp}
+}
+
+// Struct containing the mean and stdev.
+type meanStdev struct {
+	mean  float64
+	stdev float64
+}
+
+// Calculates the mean and standard deviation of the X and Z coordinates stored in the point cloud.
+func calculatePointCloudMeanAndStdevXZ(cloud pointcloud.PointCloud) (meanStdev, meanStdev, error) {
+	var X, Z []float64
+	var x, z meanStdev
+
+	cloud.Iterate(0, 0, func(pt r3.Vector, data pointcloud.Data) bool {
+		X = append(X, pt.X)
+		Z = append(Z, pt.Z)
+		return true
+	})
+
+	meanX, err := safeMath(stats.Mean(X))
+	if err != nil {
+		return x, z, errors.Wrap(err, "unable to calculate mean of X values on given point cloud")
+	}
+	x.mean = meanX
+
+	stdevX, err := safeMath(stats.StandardDeviation(X))
+	if err != nil {
+		return x, z, errors.Wrap(err, "unable to calculate stdev of Z values on given point cloud")
+	}
+	x.stdev = stdevX
+
+	meanZ, err := safeMath(stats.Mean(Z))
+	if err != nil {
+		return x, z, errors.Wrap(err, "unable to calculate mean of Z values on given point cloud")
+	}
+	z.mean = meanZ
+
+	stdevZ, err := safeMath(stats.StandardDeviation(Z))
+	if err != nil {
+		return x, z, errors.Wrap(err, "unable to calculate stdev of Z values on given point cloud")
+	}
+	z.stdev = stdevZ
+
+	return x, z, nil
+}
+
+// Calculates the scaling factor needed to fit the projected pointcloud to the desired image size, cropping it
+// if needed based on the mean and standard deviation of the X and Z coordinates.
+func calculateScaleFactor(xRange, zRange float64) float64 {
+	var scaleFactor float64
+	if xRange != 0 || zRange != 0 {
```

Comment:
Doesn't this allow for divide by zero errors, which will cause a panic?
e.g. `xRange` == 10 `zRange` == 0